### PR TITLE
Add dismax when calculating scores of synonym expansions #281

### PR DESCRIPTION
**Pull request by renekrie**
This implements #281. 

The implementation adds a new parameter `multiMatchTie`. If multiMatchTie < 1.0, a dismax query is added to produce the query structure described in the comments on issue #281. As a result, if for example, an input query `laptop` is expanded into `laptop or notebook` the dismax aggregates the scores of the two terms. If multiMatchTie=0 only the max score will be used while the 'old' `tie` param is still applied to field matches (f1:laptop | f2:laptop).

At what stage is the new dismax query added? In general, we transform queries in several steps:

`(user query string) => (Querqy object model) => (Lucene Query Factories) => (Lucene Query)`

The new dismax is added here:

`(Lucene Query Factories) -> Add dismax within Lucene Query Factory structure => (Lucene Query)`

(see querqy.lucene.rewrite.MultiMatchDismaxQueryStructurePostProcessor)

Doing it at this late stage has the advantage that we already know the Lucene structure of the query - all field names have been applied and the analysis chain will not delete any term later. Doing it within  `(Querqy object model)` wasn't possible within the current model design (a Dismax cannot be a child query of a Dismax parent). I can also imagine that we will introduce further  `PostProcessor`s, for example for optimising the query by removing duplicate tokens that where created independently by two rewriters.

I've added a visitor pattern to Lucene Query Factories for easier navigation of the query tree.

### CODE REVIEW

File: querqy-for-lucene/querqy-lucene/src/main/java/querqy/lucene/rewrite/MultiMatchDismaxQueryStructurePostProcessor.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        this.multiMatchTieBreakerMultiplier = multiMatchTieBreakerMultiplier;
+    }
+
+    /**
+     * Rewrite the query structure. The operation might manipulate the input query structure and not operate on a copy.
+     *
+     * @param structure The input query structure.
+     * @return The rewritten structure.
+     */
+    public LuceneQueryFactory<?> process(final LuceneQueryFactory<?> structure) {
```
Comment (JohannesDaniel) [inline querqy-for-lucene/querqy-lucene/src/main/java/querqy/lucene/rewrite/MultiMatchDismaxQueryStructurePostProcessor.java:52]: What can I say... 

Some references to Uncle Bob:
- functions should not be large enough to hold nested control structures (conditionals and loops)
- blocks should be one line long, consisting probably of a single function call
- an ideal function is two to four lines of code long

**Review by JohannesDaniel — APPROVED** [review]
If I understand correctly, the multiMatchTie is only applied on the top level DMQs of a query. However, I think there will be cases where synonym-related disjunctions start at a lower level, e.g.
```
iphone8 -> 
q(
  dmq(
    iphone8, 
    bool(
      dmq(iphone), 
      dmq(8)
    )
  )
)
```

What would happen if we now expand the query by the rule 
```
iphone => 
  SYNONYM: apple smartphone
```
?

In this case the tie would work like before, right?

**Comment by renekrie** [comment]
> If I understand correctly, the multiMatchTie is only applied on the top level DMQs of a query. However, I think there will be cases where synonym-related disjunctions start at a lower level, e.g.
> 
> ```
> iphone8 -> 
> q(
>   dmq(
>     iphone8, 
>     bool(
>       dmq(iphone), 
>       dmq(8)
>     )
>   )
> )
> ```
> 
> What would happen if we now expand the query by the rule
> 
> ```
> iphone => 
>   SYNONYM: apple smartphone
> ```
> 
> ?
> 
> In this case the tie would work like before, right?

That's indeed a good point. I had been thinking about it and assumed that we'd normally not apply synonyms on generated terms. But the more I look at it, the more I think it could happen (not sure but maybe the Replace Rewriter would create a Boolean Query at this level?) - I'll implement this.

File: querqy-for-lucene/querqy-lucene/src/main/java/querqy/lucene/QueryParsingController.java [hunk 2]
```diff
@@ -126,7 +127,7 @@ public QueryParsingController(final LuceneSearchEngineRequestAdapter requestAdap
             boostTermQueryBuilder = null;
             boostSearchFieldsAndBoostings = null;
             builder = new LuceneQueryBuilder(new LuceneTermQueryBuilder(), queryAnalyzer, searchFieldsAndBoosting, 1f,
-                    requestAdapter.getTermQueryCache().orElse(null));
```
[the remark this answers is not in the document]
Comment (renekrie) [reply querqy-for-lucene/querqy-lucene/src/main/java/querqy/lucene/QueryParsingController.java:129]: Agreed - but let's change this in a separate PR

**Comment by JohannesDaniel** [comment]
The replace rewriter does not expand queries, it just replaces stuff. I was rather thinking about the wordbreak rewriter. And in general, I think it is not the case that common-rules ignores generated terms. It does not apply sequence extraction across hierarchies.

**Review by JohannesDaniel — APPROVED** [review]

**Pull request merged**